### ISSUE TITLE
1.3 Beta: Global shortcut Auto Type - Original window does not regain focus in a multiple-screen setup

### ISSUE THREAD

**Issue by Loilo** [keeweb/keeweb#307]
Disclaimer: I switched to beta-mode via 1.2.4 settings, not an explicitly compiled 1.3 version. All of this happens on OS X.

Using the Ctrl + Shift + T shortcut I encountered the following issue: If the previously focused window is on a secondary screen it will only regain focus if there are no other windows of the same application.
For example if I have multiple Opera windows on multiple screens (which I do regularly) and I want to log in to a website on the second screen's browser window, KeeWeb will set focus to the browser window on my primary screen. It would probably also enter my credentials there but I can't say that for sure because #306 happens.

**Comment by antelle** [comment]
not sure if we can control it at all, needs verification on v1.3. please don't test auto-type in v1.2.4, it will not work at all

**Comment by antelle** [comment]
btw, what happens if you hide keeweb window with ⌘H? At the moment, I don't have an external display for my mac to test it...

**Comment by Loilo** [comment]
I'll look into that tomorrow. I also only got a second screen at work where I use a Cinema Display as a second screen for my MacBook.

**Comment by antelle** [comment]
So, I installed WiDisplay and verified it: after hiding KeeWeb with ⌘H, Chrome on the same desktop gets focus. At the moment I don't have a solution for this.

**Comment by antelle** [comment]
smith

**Comment by antelle** [comment]
Also within single screen, on different Desktops, the app behaves strange as well.

**Comment by braham-snyder** [comment]
Some of [kwm](https://github.com/koekeishiya/kwm)'s code could help here. Among other things, it provides commands for sending the focused window to the previously-opened space, and for opening the previously-opened space. 

At the very least, it's a nice keyboard-based, temporary workaround for this issue: note that you'll probably need to temporarily unfocus KeeWeb's password field when it pops up (using <kbd>TAB</kbd>) since said field appears to consume the relevant key-presses before kwm (well, kwm's hotkey daemon) can see them. (After moving KeeWeb to the desired space, you can refocus the password field with <kbd>SHIFT</kbd><kbd>TAB</kbd>.) Tangentially, this overzealous key-consumption behavior probably deserves its own issue (are other people encountering it?).

**Comment by braham-snyder** [comment]
Instead of the above, "Options" > "Assign To" > "All Desktops" (right click KeeWeb in your dock) is actually working perfectly for me (I thought I'd tried this before using kwm, but after a few days I'm not seeing any problems whatsoever).

**Comment by Raptor399** [comment]
> "Options" > "Assign To" > "All Desktops" (right click KeeWeb in your dock) is actually working perfectly for me

Alas, that trick isn't working here (KeeWeb 1.5.4).

In my experience Auto Type only works correctly 100% of the time when either:

* all browser windows are on the same desktop
* KeeWeb pops up on the same desktop as the browser window
* KeeWeb was already unlocked

This makes working with KeeWeb quite tricky in a three monitor setup.

Would it be possible to validate browser information before actually pasting in the info - to prevent unintended pasting?

**Comment by antelle** [comment]
Yes, browser information will be validated in future versions, however multi-screen issues are tricky to solve.

**Comment by braham-snyder** [comment]
just to clarify, I'm indeed seeing the issue @Raptor399 detailed (just took a while for me to encounter the scenario he described) -- I've edited my prior post to reflect this

**Comment by Raptor399** [comment]
Seems there are two separate issues:
* Pasting information into the wrong browser window
* Focussing the correct browser window to paste info into

The first issue is critical and luckily a future version will solve it, awesome!

The second issue is tricky to solve. But once the first is solved, we automatically get a workaround for the second: 

* Use Auto Type
* KeeWeb will pop up and ask for a password.
* Next, nothing will happen because validation prevents pasting.
* Use Auto Type again. This time it will always work because KeeWeb was already unlocked.

Seems like a perfectly acceptable workaround to me.

**Comment by Alph2** [comment]
Just throwing an idea out there: you could make keeweb running as a menubar app, so that it would alway be in the right space/desktop. (And it would make sense, since it works as a background app)

**Comment by mkarnicki** [comment]
The same happens on OS X if you have your browser full-screen, and KeeWeb pops up with a question to pick an entry (if auto-type couldn't figure it out). Then once KeeWeb minimizes, the autotype will happen in the primary "desktop/space", to whatever is focused, and not the browser that's full-screened to the right (on OS X). I'm not sure this is worth a separate bug report, because to me it looks like it's the same issue here.

The problem with @Alph2 's suggestion is that even if the app was in the menu bar, if you open the item (say, KeeWeb pops up to ask to select entry, if auto-type couldn't determine it), regardless it **will** again go back to your primary space, and then what I just described above happens anyways.

**Comment by andremann** [comment]
It looks like this is working in v1.16.1 (2020-12-04)!!

**Comment by robin-moser** [comment]
Unfortunately I still have these problems, v1.16.1 on MacOS Big Sur.
I have my Macbook setup with two dual Monitors and the lid closed. 

Let's assume, Keeweb runs on the second Monitor in Fullscreen. Autotype from an app on the fist monitor it working, but the second Monitor switches to Keeweb and stays there. If I want to perform Autotype on the second monitor, keeweb has to be on the same "desktop", so neither the app i want to type in nor Keeweb are able to be on fullscreen, and also not on two different desktops.

**Comment by andremann** [comment]
Not sure what happened, but it doesn't work anymore as I described one week ago... sadly...
